internal/rsl/rsl.go

```diff
@@ -513,6 +514,47 @@ func GetFirstReferenceEntryForRef(repo *git.Repository, targetRef string) (*Refe
 	return firstEntry, annotations, nil
 }
 
+// SkipAllInvalidReferenceEntries skips all reference entries that could have previously been
+// normal reference entries, but after a force change a ref that the rsl entries point to, some of the
+// entries may point to non-existent commits. This function adds an annotation that skips all entries
+// that are pointing to non-existent commits
+func SkipAllInvalidReferenceEntries(repo *git.Repository, signCommit bool) error {
+	slog.Debug("Checking if RSL entries point to non-existent commits...")
+
+	iterator, err := GetLatestEntry(repo)
+	if err != nil {
+		return err
+	}
+
+	entriesToSkip := []plumbing.Hash{}
+
+	for {
+		if entry, ok := iterator.(*ReferenceEntry); ok {
+			if _, err := gitinterface.GetCommit(repo, entry.TargetID); err != nil {
+				if errors.Is(err, plumbing.ErrObjectNotFound) {
```

Comment:
This isn't right when invoked on the same copy of the repository where the rebase is performed; a rebase doesn't mean the commit isn't in the object store anymore. The commit sticks around until the garbage collector cleans it up.

Instead, we want to use `KnowsCommit` from the current commit. By making this _per ref_, we walk back until we find a reference entry for the ref whose target is known by the current commit. Take the following example:

A <- B <- C <- D <- E | This is the original commit history for a branch, we have RSL entries at A, B, D, and E
A <- B <- F                 | This is the new commit history for the branch, C, D, and E are squashed to create F

 We walk back until we find the RSL entry for B which is _known_ by F. Clearly, the entry for D and E are invalid now. Thus, the function:
 a) creates an annotation entry skipping the RSL entries at D and E
b) creates a reference entry at F

Does this make sense? :smile:

Reply:
cc @wlynch and @justincappos for if this workflow makes sense.

Reply:
I have changed the logic to reflect your idea. Was this what you meant?